File: ios/Plugin/Logger.swift
//
//  Logger.swift
//  WillsubCapacitorSplashscreen
//
//  Created by Aparajita on 10/13/20.
//

import Capacitor

class Logger {
  var level: LogLevel = .info

  // Log levels from lowest to highest. Log messages at or below the current log level
  // are displayed.
  enum LogLevel: Int {
    case off
    case error
    case warn
    case info
    case debug
    case trace
  }

  func setLogLevel(_ logLevel: String) {
    switch logLevel.lowercased() {
    case "trace":
      level = .trace

    case "debug":
      level = .debug

    case "info":
      level = .info

    case "warn":
      level = .warn

    case "error":
      level = .error

    case "off":
      level = .off

    default:
      level = .info
    }
  }

  private func canLog(withLevel level: LogLevel) -> Bool {
    return self.level != .off && self.level.rawValue >= level.rawValue
  }

  /*
   * Print a message to the log.
   */
  func log(_ prefix: String, _ items: [Any]) {
    let message = items.map { "\($0)" }.joined(separator: " ")
    CAPLog.print("\(prefix) [\(String(describing: WSSplashScreen.self))]", message)
  }

  /*
   * Print a trace message to the log.
   */
  func trace(_ items: Any...) {
    if canLog(withLevel: .trace) {
      log("🔎", items)
    }
  }

  /*
   * Print a debug message to the log.
   */
  func debug(_ items: Any...) {
    if canLog(withLevel: .debug) {
      log("👉", items)
    }
  }

  /*
   * Print an info message to the log.
   */
  func info(_ items: Any...) {
    if canLog(withLevel: .info) {
      log("🟢", items)
    }
  }

  /*
   * Print a warning message to the log.
   */
  func warn(_ items: Any...) {
    if canLog(withLevel: .warn) {
      log("🟠", items)
    }
  }

  /*
   * Print an error message to the log.
   */
  func error(_ items: Any...) {
    if canLog(withLevel: .error) {
      log("🔴", items)
    }
  }
}

File: android/src/main/java/com/aparajita/capacitor/splashscreen/Logger.java
package com.aparajita.capacitor.splashscreen;

import android.util.Log;
import java.util.HashMap;

public class Logger {

    enum LogLevel {
        off,
        error,
        warn,
        info,
        debug,
        trace
    }

    static HashMap<LogLevel, String> prefixes;

    static {
        prefixes = new HashMap<>();
        prefixes.put(LogLevel.error, "🔴");
        prefixes.put(LogLevel.warn, "🟠");
        prefixes.put(LogLevel.info, "🟢");
        prefixes.put(LogLevel.debug, "👉");
        prefixes.put(LogLevel.trace, "🔎");
    }

    LogLevel currentLevel = LogLevel.info;

    public void setLogLevel(LogLevel logLevel) {
        currentLevel = logLevel;
    }

    public void setLogLevel(String logLevel) {
        try {
            currentLevel = LogLevel.valueOf(logLevel.toLowerCase());
        } catch (IllegalArgumentException ex) {
            // ignore
        }
    }

    /**
     * Print a trace message to the log.
     */
    public void trace(String module, String... items) {
        print(LogLevel.trace, module, items);
    }

    /**
     * Print a debug message to the log.
     */
    public void debug(String module, String... items) {
        print(LogLevel.debug, module, items);
    }

    /**
     * Print an info message to the log.
     */
    public void info(String module, String... items) {
        print(LogLevel.info, module, items);
    }

    /**
     * Print a warning message to the log.
     */
    public void warn(String module, String... items) {
        print(LogLevel.warn, module, items);
    }

    /**
     * Print an error message to the log.
     */
    public void error(String module, String... items) {
        print(LogLevel.error, module, items);
    }

    /*
     * Print a message to the log.
     */
    private void print(LogLevel level, String module, String[] items) {
        if (!canLogAt(level)) {
            return;
        }

        StringBuilder message = new StringBuilder();

        for (int i = 0; i < items.length; i++) {
            if (i == 0) {
                message.append(items[0]);
            } else {
                message.append(" ").append(items[i]);
            }
        }

        String prefix = prefixes.get(level);
        String msg = String.format("%s [%s] %s", prefix, module, message);

        switch (level) {
            case info:
                Log.i(module, msg);
                break;
            case warn:
                Log.w(module, msg);
                break;
            case error:
                Log.e(module, msg);
                break;
            case trace:
                Log.v(module, msg);
                break;
        }
    }

    private boolean canLogAt(LogLevel level) {
        return this.currentLevel != LogLevel.off && this.currentLevel.compareTo(level) >= 0;
    }
}
